### PR TITLE
Automatically generate `filterset_class` for snippets and generic `IndexView` through `list_filter`

### PR DESCRIPTION
**Pull request by laymonage**
As part of wagtail/rfcs#85.

This PR adds a `get_filterset_class` on the generic `IndexView` that automatically generates a `WagtailFilterSet` subclass with the fields defined in `IndexView.list_filter`. The API is also surfaced in the `SnippetViewSet` class to ease customisation for Snippets' `IndexView`. 

For Snippets' `IndexView`, the `get_filterset_class()` method works by calling `SnippetViewSet.get_filterset_class()` first. If the result is `None`, it then falls back to the generic `IndexView.get_filterset_class()` implementation where the class is dynamically generated if necessary.

_Please check the following:_

-   [x] Do the tests still pass?[^1]
-   [x] Does the code comply with the style guide?
    -   [x] Run `make lint` from the Wagtail root.
-   [x] For Python changes: Have you added tests to cover the new/fixed behaviour?
-   [x] For new features: Has the documentation been updated accordingly?

**Please describe additional details for testing this change**.

Continuing from the test example in #10241.

```py
from wagtail.snippets.models import register_snippet
from wagtail.snippets.views.snippets import SnippetViewSet


class PersonViewSet(SnippetViewSet):
    icon = "user"
    list_per_page = 2
    chooser_per_page = 1
    admin_url_namespace = "some_namespace"
    base_url_path = "deep/within/the/admin"
    chooser_admin_url_namespace = "my_chooser_namespace"
    chooser_base_url_path = "choose/wisely"
    # this will generate exact filters
    list_filter = ["job_title", "first_name"]
    # alternatively, this will generate filters with lookups
    list_filter = {
        "job_title": ["contains"],
        "first_name": ["contains"],
        "first_published_at": ["date"],
    }



register_snippet(Person, viewset=PersonViewSet)
```

[^1]: [Development Testing](https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
[^2]: [Browser and device support](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)
[^3]: [Accessibility Target](https://docs.wagtail.org/en/latest/contributing/developing.html#accessibility-targets)

### CODE REVIEW

**Comment by laymonage** [comment]
Hold on, I haven't added docs in the `topics/snippets.md` file.

**Review by gasman — APPROVED** [review]
All looks good to me!

File: wagtail/test/testapp/migrations/0023_fullfeaturedsnippet_country_code_and_more.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
+# Generated by Django 4.2b1 on 2023-03-21 17:12
```
Comment (gasman) [inline wagtail/test/testapp/migrations/0023_fullfeaturedsnippet_country_code_and_more.py:1]: Looks like this file needs renumbering - will do that while I merge.

**Comment by gasman** [comment]
Merged in a3498b260f8d2ed26ebd3df865e32fd84669d2dd

**Pull request closed without merging**